pkg/operator/vsphereproblemdetector/vsphere_problem_detector_deployment.go

```diff
@@ -0,0 +1,189 @@
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				c.versionGetter.SetVersion(deploymentControllerName, c.targetVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "WaitDeployment"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
+
+	resourcemerge.SetDeploymentGeneration(&opStatus.Generations, deployment)
+
+	updateGenerationFn := func(newStatus *operatorapi.OperatorStatus) error {
+		if deployment != nil {
+			resourcemerge.SetDeploymentGeneration(&newStatus.Generations, deployment)
+		}
+		return nil
+	}
+
+	if _, _, err := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(deploymentAvailable),
+		v1helpers.UpdateConditionFn(deploymentProgressing),
+		updateGenerationFn,
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *VSphereProblemDetectorDeploymentController) Name() string {
+	return deploymentControllerName
+}
+
+func (c *VSphereProblemDetectorDeploymentController) Run(ctx context.Context, workers int) {
+	// This adds event handlers to informers.
+	ctrl := c.factory.WithSync(c.Sync).ToController(deploymentControllerName, c.eventRecorder)
+	ctrl.Run(ctx, workers)
+}
+
+func (c *VSphereProblemDetectorDeploymentController) getRequiredDeployment(spec *operatorapi.OperatorSpec) *appsv1.Deployment {
+	deploymentString := string(generated.MustAsset("vsphere_problem_detector/06_deployment.yaml"))
+
+	pairs := []string{
+		"${OPERATOR_IMAGE}", os.Getenv(vSphereProblemDetectorOperatorImage),
+	}
+
+	replacer := strings.NewReplacer(pairs...)
+	deploymentString = replacer.Replace(deploymentString)
+	// Replace log level
+	logLevel := getLogLevel(spec.LogLevel)
+	deploymentString = strings.ReplaceAll(deploymentString, "${LOG_LEVEL}", strconv.Itoa(logLevel))
+
+	deployment := resourceread.ReadDeploymentV1OrDie([]byte(deploymentString))
+	return deployment
+}
+
+// factory.PostStartHook to poke newly started controller to resync.
+// This is useful if a controller is started later than at CSO startup
+// - CSO's CR may have been already processes and there may be no
+// event pending in its informers.
+func initalSync(ctx context.Context, syncContext factory.SyncContext) error {
+	syncContext.Queue().Add(factory.DefaultQueueKey)
+	return nil
+}
+
+func getLogLevel(logLevel operatorapi.LogLevel) int {
```

Comment:
Please move it to a common package, it's used in deploymentcontroller.go too.

Reply:
fixed. I extracted most of deployment controller creation code in a util function and used in both places.